lnwallet/wallet.go

```diff
@@ -1392,19 +1436,27 @@ func selectInputs(amt btcutil.Amount, coins []*Utxo) (btcutil.Amount, []*Utxo, e
 }
 
 // coinSelect attempts to select a sufficient amount of coins, including a
-// change output to fund amt satoshis, adhering to the specified fee rate. The
-// specified fee rate should be expressed in sat/kw for coin selection to
-// function properly.
-func coinSelect(feeRate SatPerKWeight, amt btcutil.Amount,
-	coins []*Utxo) ([]*Utxo, btcutil.Amount, error) {
+// change output to fund amt satoshis, adhering to the specified fee rate. It
+// returns the size of the created funding ouput, as well as the size of the
+// change output if any. The specified fee rate should be expressed in sat/kw
+// for coin selection to function properly, and the passed dustLimit is used to
+// clamp the outputs created.
+//
+// If subtractFees is true, then the estimated fee will be taken from the
+// output amount, meaning that the resuling funding output might be smaller
+// than the passed amt. The resulting funding output value will be returned,
+// along with the change amount if any.
+func coinSelect(feeRate SatPerKWeight, amt, dustLimit btcutil.Amount,
```

Comment:
It might just be me, but I find this method a bit difficult to follow in its final form due to all of the different variables involved, though avoiding the code duplication is certainly nice.

Reply:
Yeah, I agree. In the process of trying to share the code between the two cases, it kinda got out of hand. The two cases (`subtractFee=true/false`) are different enough to make it tricky to handle edge cases between them.

Code harder to reason about is more prone for errors than a little bit of code duplication.

Reply:
I created a version where the two cases are separated, lmk what you think: https://github.com/lightningnetwork/lnd/compare/master...halseth:autopilot-spend-coins-subtract-fees-distinct-coinselect?expand=1

Reply:
Separating them LGTM. Would also be nice to get someone else's opinion.